File: C-Plus-Plus/dp/golomb_sequence.cpp
/* C++ program to print the n'th term in the Golomb sequence
Golomb sequence is a non-decreasing integer sequence where n'th
term is  equal to the number of times n appears in the sequence */

#include <bits/stdc++.h>
using namespace std;

int golomb_sequence(int n)
{
    // Create a dp array, with value initialized as 0.
    int dp[n + 1];

    dp[1] = 1;

    // Identify the previous term 'prev' and go prev terms behind and find a number.
    // Now assign the current element with an incremented value of that element.
    for (int i = 2; i <= n; i++)
    {
        int prev = dp[i - 1];
        int back_index = i - dp[prev];
        dp[i] = 1 + dp[back_index];
    }
    return dp[n];
}

int main()
{
    int n;
    cout << "Enter the value of n?, where you need the n'th number in the golomb sequence.";
    cin >> n;
    if (n <= 0)
    {
        cout << "The given value of n is invalid.";
        return 0;
    }
    int res = golomb_sequence(n);
    cout << "The " << n << "'th term in the golomb sequence is " << res;
    return 0;
}

/*
Time Complexity: O(n), where 'n' is the given number
Space Complexity: O(n)

SAMPLE INPUT AND OUTPUT

SAMPLE 1

Enter the value of n?, where you need the n'th number in the golomb sequence. 5
The 5'th term in the golomb sequence is 3.

SAMPLE 2

Enter the value of n?, where you need the n'th number in the golomb sequence. 867
The 867'th term in the golomb sequence is 79.
*/

File: Dart/dp/GolombSequence.dart
/*
Golomb sequence is a non-decreasing integer sequence where n'th
term is  equal to the number of times n appears in the sequence 
*/

import 'dart:io';

// Recursive golomb function
int golomb(int n){

    //Base case
    if(n==1){
      return 1;
    }
    // recursive case
    return 1 + golomb(n -golomb(golomb(n - 1)));
}

// Function to add golomb sequence in a list
int print_golomb(int n){
  List<int> dp = [];

  // Add golomb numbers to the list
  for(int i=1; i<n+1; i++){
    dp.add(golomb(i));
  }

  // Print list element at index n
  return dp.elementAt(n-1);
}

void main(){
  print("Enter the value of 'n', where you need the n'th number in the golomb sequence:");
  int num = int.parse(stdin.readLineSync()!);

  if (num <= 0){
    print("The given value of n is invalid.");
  }

  // pass input to golomb calculator
  int res = print_golomb(num);

  print("The ${num}th term in the golomb sequence is $res");
}

/*
Time Complexity: O(n), where 'n' is the given number
Space Complexity: O(n)

Sample I/O:

Enter the value of 'n', where you need the n'th number in the golomb sequence:
12
The 12th term in the golomb sequence is 6

Enter the value of 'n', where you need the n'th number in the golomb sequence:
30
The 30th term in the golomb sequence is 10

Enter the value of 'n', where you need the n'th number in the golomb sequence:
60
The 60th term in the golomb sequence is 15
*/

